general improvements

TheCataliasTNT2k opened the pull request:
This PR adds some general improvements:
- when using UID with uppercase enabled, look-alike characters are excluded (e.g. lowercase L, uppercase I may look the same)
- The generated short link is now shown completely, not only the short section
- Some styles have been changed to align items vertically
- docker build is now possible for me, without pulling the base image before manually
- the project does no longer contain inline js, which is blocked by some browsers and my reverse proxy for security reasons to prevent XSS (i.e. `content-security-policy
script-src 'self'; object-src 'none'; base-uri 'none';`

SinTan1729 commented:
Hi. Thanks for the PR. I've taken a quick look, and everything seems fine except for the large commit fb6610a244658bc195350bc48ed21ce5ceb2dd1b.

I'd definitely need to take more time before merging anything, but I recommend that you remove that commit from this PR, split it up into smaller commits with a single purpose for each, and open another PR.

Some comments about that commit:
1. I don't see how `open_url` is useful. If clicking on the shorturl opens it, it'll immediately be handled by the backend, so essentially it'll be the same as clicking on the longurl. That's already possible.
2. I agree with getting rid of inline js, but there's some issue with one part where we need to add `/` at the end of the address. Otherwise, the images and scripts don't work reliably with a custom landing directory.
3. I have some concerns about the changed flow of some of the paths, but I can take a look at it again if you make a new PR.

SinTan1729 commented:
Also, in 6fceec60abc1511ee6413cf59066f786ed4fe1e5, why did you remove `i`, `j`? I don't see how they're ambiguous.

TheCataliasTNT2k commented:
> Also, in [6fceec6](https://github.com/SinTan1729/chhoto-url/commit/6fceec60abc1511ee6413cf59066f786ed4fe1e5), why did you remove `i` and `j`? I don't see how they're ambiguous.

Because they look very similar, like uppercase o and 0.
Yes, the problem with I and l is bigger, but I chose to remove them anyway.
We can add them back, if you want.

SinTan1729 commented:
I think those should be added back.

TheCataliasTNT2k commented:
> Yeah, I think those should be added back.

But removing O and 0 is ok then?

SinTan1729 commented:
Yes, I think that's fair. They look very similar in some fonts.

TheCataliasTNT2k commented:
Ok, did that, I will push the commits
Now we need to resolve the things with the big one.

SinTan1729 commented:
> Well, it can be deactivated (and is by default).
> However, we can remove this, if it is a problem.

I think it should be removed, since it simply doesn't serve any purpose.

> Then you might want to address this issue in the backend, not in the frontend.
> Or we move it into the script.js file.
> This script block will not be executed by my browser, and might not be for other users.

I don't think it can be handled in the backend due to the idiosyncrasies of HTML. But it might be possible to put it in `script.js` and assign it as an `.onload` function.

> What do you mean with that?
> And what should I put in a new PR?

I just mean I need to do more testing, since the flows are quite different. Don't worry about it for now.

> I tried to split up the commit, however I could not get git to accept only the things I want to, so I needed to commit it as one.
> I initially did not have all of these changes at the same time, but because of broken stuff, I needed to do it like that.

I understand but it's pretty much incomprehensible due to the many changes that are happening simultaneously.

TheCataliasTNT2k commented:
> I think it should be removed, since it simply doesn't serve any purpose.

Done.

> I don't think it can be handled in the backend due to the idiosyncrasies of HTML. But it might be possible to put it in `script.js` and assign it as an `.onload` function.

Can you tell me exactly, why this js code existed? Maybe I find another way to do this, without breaking the user's history and executing this on load.

> I just mean I need to do more testing, since the flows are quite different. Don't worry about it for now.

Actually, the flows are exactly the same, just in one function, which handles both, instead of two functions.
The called function determines, if the user wants to login or logout and calls the other functions, which were called directly before.

> I understand but it's pretty much incomprehensible due to the many changes that are happening simultaneously.

There are three things happening:
- vars are now const
- I removed the inline js stuff
- I added `await` statements at every needed point, to force execution in the right order.

SinTan1729 commented:
> Can you tell me exactly, why this js code existed?

When the `/admin/manage` path is accessed, depending on whether there is a trailing slash, the resources will be requests from different addresses. It happens due to how relative links are processed. If there is no trailing slash, they will be requested from `/admin/*`, and from `/admin/manage/*` otherwise.

Since they are part of the HTML, I can't use javascript to fix them either.

SinTan1729 commented:
> Actually, the flows are exactly the same, just in one function, which handles both, instead of two functions.
> The called function determines, if the user wants to login or logout and calls the other functions, which were called directly before.

Ok, I'll still prefer to deal with it as a separate PR. How about this? You create a new branch, cherry pick this commit there, and open a PR. And then remove it from here. I'll deal with any merge conflicts.

TheCataliasTNT2k commented:
> > Actually, the flows are exactly the same, just in one function, which handles both, instead of two functions.
> > The called function determines, if the user wants to login or logout and calls the other functions, which were called directly before.
> 
> Ok, I'll still prefer to deal with it as a separate PR. How about this? You create a new branch, cherry pick this commit there, and open a PR. And then remove it from here. I'll deal with any merge conflicts.

Well, I then will need to try to split up the one commit. Give me some time.

TheCataliasTNT2k commented:
> > Can you tell me exactly, why this js code existed?
> 
> When the `/admin/manage` path is accessed, depending on whether there is a trailing slash, the resources will be requests from different addresses. It happens due to how relative links are processed. If there is no trailing slash, they will be requested from `/admin/*`, and from `/admin/manage/*` otherwise.
> 
> Since they are part of the HTML, I can't use javascript to fix them.

This route exists in the backend, but nowhere in your frontend.
According to your docker-compose, you need to replace ALL files, if you want to use this, so the missing script does not matter (the index.html would be replaced).
It does make sense, that you might need it, when using a custom frontend.

SinTan1729 commented:
> This is just normally not the way to do it, even in the script file.

Yes, it does look weird. I think I can invoke a redirect from the server, which seems to be the standard way to do this. Let's deal with this in the other PR.
